### PR TITLE
K3s VMware Setup

### PR DESCRIPTION
**Pull request by Hareet**
k3s cluster setup on a VMware datacenter that uses VMware out-of-tree Cloud Provider Interface and Container Storage Interface. 

This enables the cluster to auto-provision storage volume mounts using CSI that talks to VMware API. Allows k3s to orchestrate pvc with ability to failover to a spare node if `drain node` is invoked.

### CODE REVIEW

File: content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+Take a peak at all 3 vsphere-controller-manager pods logs to ensure nothing is immediately erring. Common errors are using the incorrect datacenter name, UUIDs for VMs in the k3s curl command, or invalid credentials in the configmap and secrets resources created in step 2 above. If one of these errors is displaying in the log, you will want to delete the deployment (in step 3 above, replace `apply` with `delete`, edit the yaml and re-deploy (run step 3 again).
+
+
+## Deploy VMware Container Storage Interface (CSI) to your k3s cluster
+
+[VMware documentation for CSI](https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/2.0/vmware-vsphere-csp-getting-started/GUID-A1982536-F741-4614-A6F2-ADEE21AA4588.html)
+
+We'll be following the guide linked above.
+
+Procedure:
```
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md:295]:
```suggestion
Follow the [VMware documentation for CSI](https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/2.0/vmware-vsphere-csp-getting-started/GUID-A1982536-F741-4614-A6F2-ADEE21AA4588.html) with these steps:
```

File: content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+cluster-id = "<cluster-id>"
+
+[VirtualCenter "<IP or FQDN>"]
+insecure-flag = "<true or false>"
+user = "<username>"
+password = "<password>"
+port = "<port>"
+datacenters = "<datacenter1-path>, <datacenter2-path>, ..."
+```
+
```
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md:326]:
```suggestion
3) Create kubernetes secret, which will map authentication credentials and datacenter name to CSI containers. First, create a file  `/etc/kubernetes/csi-vsphere.conf`. Be sure to replace  `<vCenter_IP>`, `<USERNAME>`, `<PASSWORD>` ,  `<true_or_false>`,  `<PORT>` ,  `<datacenter1-path>` and  `<datacenter1-path>`  with your values:

   
   [Global]
   cluster-id = "<cluster-id>"
   
   [VirtualCenter "<vCenter_IP>"]
   insecure-flag = "<true_or_false>"
   user = "<USERNAME>"
   password = "<PASSWORD>"
   port = "<PORT>"
   datacenters = "<datacenter1-path>, <datacenter2-path>, ..."


```

File: content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+There is one minor edit, typically found on line 217-218, under the deployment specification for vsphere-csi-controller. 
+
+Before edit (original value)
+```
+      nodeSelector:
+        node-role.kubernetes.io/control-plane: ""
+```
+
+Please add "true" as the value for this key, seen below:
```
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md:342]:
```suggestion
Please add `true` as the value for this key, seen below:
```

File: content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+```
+/usr/local/bin/k3s kubectl -n vmware-system-csi apply -f vsphere-csi-driver.yaml
+```
+
+Follow the [verification steps seen here in Step 2 of Procedure](https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/2.0/vmware-vsphere-csp-getting-started/GUID-54BB79D2-B13F-4673-8CC2-63A772D17B3C.html)
+
+
+### Create StorageClass in k3s cluster
+
+We'll need to create a global [StorageClass](https://kubernetes.io/docs/concepts/storage/storage-classes/) resource in our k3s cluster, so cht-core deployments will be able to ask for persistent storage volumes from the k3s cluster.
```
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md:358]:
```suggestion
We'll need to create a global [StorageClass](https://kubernetes.io/docs/concepts/storage/storage-classes/) resource in our k3s cluster, so CHT Core deployments will be able to ask for persistent storage volumes from the k3s cluster.
```

File: content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+```
+/usr/local/bin/k3s kubectl apply -f vmware-storageclass.yaml
+```
+
+## Deploying a CHT-Core Project to your new k3s Cluster running on VMware
+
+This step will neatly fit into helm chart configurations, but here are the manual steps for time being.
+
+In your PVC template (peristent volume), here is a snippet of what it should resemble for all CouchDB's:
+* Note the storageClassName parameter should be identical to the storageClass we deployed earlier
```
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md:383]:
```suggestion
Your persistent volume (PVC) template for all CouchDB's should be as shown below. Note the `storageClassName` parameter should be identical to the `storageClass` we deployed earlier:
```

File: content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+* [Persistent Volume Claim](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) - This is where our project data will be stored. Important to ensure you have configured this correctly, with retain policies intact so the data is not deleted if the project is removed. It's also vital to ensure you have a backup policy either set-up in VMware vCenter GUI or you have configured the csi-snapshotter that comes with vSphere CSI.
+* [Services](https://kubernetes.io/docs/concepts/services-networking/service/) - This is utilized for CouchDB nodes to discover each other through DNS rather than internal IPs, which can change. This is also used in the COUCH_URL so API containers can discover where CouchDB is running.
+
+
+## k3s Troubleshooting
+Another doc to be linked here
```
Comment (mrjones-plip) [inline content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md:414]: We can add this content when we have it, until then, we'll just remove the place holder.

```suggestion
```

**Review by mrjones-plip — CHANGES_REQUESTED** [review]
Wow WOW!   What an epic amount of work - congrats!!!

I went ahead and moved the file up a directory and renamed it to match our existing docs.  As well, I'm only reviewing this for document structure, slightly reducing the wall of text feeling and parity on code snippets all using same nomenclature ("password" vs password vs "`PASSWORD`" vs credentials etc). I'm not at all reviewing this for technical accuracy in deploying k3s!

finally, it's impossible (as far as I know) to do code blocks inside code blocks for suggestions, so I fell back to triple (`   `) or sextuple (`      `) spaces which achieves the same thing in the final markdown - whew!

File: content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+---
+
+{{% pageinfo %}}
+This page details initial k3s cluster setup on a VMware datacenter running vSphere 7+. After initial completion, the administrator should only be adding VMs to the cluster or deploying cht-core projects to be orchestrated.
+
+{{% /pageinfo %}}
+
+### About container orchestration
+
+A container orchestrater has the ability to schedule projects on available resources spread across a datacenter. For national scale projects, or a large number of cht-core deployments to maintain, we recommend and support using a lightweight kubernetes orchestrator called [k3s](https://docs.k3s.io/). The orchestrator will monitor resources across a group of nodes (VMs), place cht-core projects where there is available resource and migrate projects to spare resources if combined utilization is high or there are underlying issues. Instead of provisioning one VM per cht-core project, we will provision larger VMs and deploy multiple cht-core projects on one VM, with each project receiving optional resource limitations.
```
[the remark this answers is not in the document]
Comment (Hareet) [reply content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md:16]: yup, [kube-scheduler](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/). I think you've defined it 👍

File: content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+  - CONTROL_PLANE_IP: The IP of one of the control plane servers you set up above
+  - VM_UUID: This is the UUID of this VM that we are adding as an agent/worker server in our k3s cluster
+
+```
+curl -sfL https://get.k3s.io | K3S_KUBECONFIG_MODE="644" INSTALL_K3S_EXEC="agent" INSTALL_K3S_VERSION="v1.25.14+k3s1" sh -s - \
+--docker --token <TOKEN> \
+--server https://<CONTROL_PLANE_IP>:6443 \
+--kubelet-arg="cloud-provider=external" \
+--kubelet-arg="provider-id=vsphere://<VM_UUID>"
+```
```
[the remark this answers is not in the document]
Comment (Hareet) [reply content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md:224]: @mrjones-plip I wasn't sure about this edit. The same formatting is applied for all the varying curl commands (a few lines above and below the optional paramter changes here).

File: content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+
+## Deploy VMware Cloud Provisioner Interface (CPI) to your k3s cluster
+
+SSH into one of your control plane servers. 
+Download the template for CPI:
+```
+VERSION=1.25
+wget https://raw.githubusercontent.com/kubernetes/cloud-provider-vsphere/release-1.25/releases/v1.25/vsphere-cloud-controller-manager.yaml
```
[the remark this answers is not in the document]
Comment (nydr) [reply content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md:232]: IMHO to realistically complete this process the reader is expected to understand standard *nix file management. Might be worth adding technical prerequisites at the top, but I don't have strong opinions regarding this.

File: content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+   
+   govc ls /
+        <datacenter-name>/vm \
+        <datacenter-name>/network \
+        <datacenter-name>/host \
+        <datacenter-name>/datastore 
+   
+   #To retrieve all Node VMs
+   govc ls /<datacenter-name>/vm \
+       <datacenter-name>/vm/<vm-name1> \
```
Comment (nydr) [inline content/en/apps/guides/hosting/4.x/self-hosting-k3s-multinode.md:157]: The tags disappears for me in the preview, does it look correctly in when rendered by the static site generator?

Minor: I'd prefer to wrap all code lines in backticks (replace `shell` with the language that is most appropriate or leave blank)
````
```shell
code goes here...
```
````
 or for shorter sections
`` `code goes here...` ``

**Review by nydr — COMMENTED** [review]
Great job summarizing the process! Even though there's some things that are guaranteed to become outdated such as version numbers I think it's a very useful reference to have.

---
@mrjones-plip already came with great feedback on the texts so I don't have much to add, mostly minor comments 

1. Path should be changed to `content/en/apps/guides/hosting/4.x/self-hosting/<...>` due to rename in #1191 

2. Should this page be specific to VMware? If so I'd prefer to rename the url (filename) to include VMware and break out `Kubernetes Concepts` into a separate page.

3. If it'll be generic down the line I'd prefer to preemptively see any VMware specific steps under [`tabpane`](https://www.docsy.dev/docs/adding-content/shortcodes/#tabbed-panes) tabs like [here](https://docs.communityhealthtoolkit.org/apps/tutorials/local-setup/#installing-docker)

**Comment by nydr** [comment]
> to do code blocks inside code blocks for suggestions

@mrjones-plip you can use more backticks
`````
````
```
hello
```
````
`````

will render as

````
```
hello
```
````

**Comment by mrjones-plip** [comment]
> @mrjones-plip Not sure if it works for suggestions but in comments you can use more backticks

OMG!!! YOU'RE MY SAVIOR.

Jokes aside, this is legit helpful. I looked and looked for an answer - I just needed to ask you!

**Review by mrjones-plip — APPROVED** [review]
Nice work @Hareet ! It ended up being easier to fix some of the code formatting and indentations myself, which I did in  f5a1d4e .

Otherwise, good to go!

**Review by Reagan — APPROVED** [review]

**Review by esthermmoturi — APPROVED** [review]

**Pull request merged**